### PR TITLE
Added solution

### PR DESCRIPTION
**Pull request by AndrewPolishchuc**

### CODE REVIEW

File: src/main/java/core/basesyntax/dao/animal/AnimalDaoImpl.java [hunk 1]
```diff
@@ -12,11 +14,33 @@ public AnimalDaoImpl(SessionFactory sessionFactory) {
...
+            }
+        }
     }
 
     @Override
     public List<Animal> findByNameFirstLetter(Character character) {
-        return null;
+        try (Session session = sessionFactory.openSession()) {
+            return session.createQuery("from Animal"
+                    + "where name like :name", Animal.class)
```
Comment (DariiaPikul) [inline src/main/java/core/basesyntax/dao/animal/AnimalDaoImpl.java:41]: I received a comment about using LIKE with the matching pattern only
Not a rule, more of a suggestion
So let's replace it with "="

**Review by DariiaPikul — COMMENTED** [review]
Looks good to me, but I might've missed something because am not practically experienced enough in the question

File: src/main/java/core/basesyntax/dao/animal/AnimalDaoImpl.java (hunk 1 above)
Comment (AndrewPolishchuc) [reply src/main/java/core/basesyntax/dao/animal/AnimalDaoImpl.java:41]: We use% to search for a word, we can only do this with the LIKE operator

File: src/main/java/core/basesyntax/dao/ma/MentorDaoImpl.java [hunk 2]
```diff
@@ -2,6 +2,7 @@
 
 import core.basesyntax.model.ma.Mentor;
 import java.util.List;
+import org.hibernate.Session;
 import org.hibernate.SessionFactory;
 
 public class MentorDaoImpl extends PersonDaoImpl implements MentorDao {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/dao/ma/MentorDaoImpl.java:8]: how about refactoring save() method in PersonDaoImpl using transaction?

File: src/main/java/core/basesyntax/dao/machine/MachineDaoImpl.java [hunk 3]
```diff
@@ -12,11 +15,33 @@ protected MachineDaoImpl(SessionFactory sessionFactory) {
...
+            }
+        }
     }
 
     @Override
     public List<Machine> findByAgeOlderThan(int age) {
-        return null;
+        try (Session session = sessionFactory.openSession()) {
+            return session.createQuery("from Machine"
+                    + " where year > :year", Machine.class)
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/dao/machine/MachineDaoImpl.java:42]:
```suggestion
                    + " where year < :year", Machine.class)
```

File: src/main/java/core/basesyntax/model/ma/Coach.java [hunk 4]
```diff
@@ -1,5 +1,10 @@
 package core.basesyntax.model.ma;
 
+import javax.persistence.Entity;
+import javax.persistence.PrimaryKeyJoinColumn;
+
+@Entity
+@PrimaryKeyJoinColumn(name = "coachId")
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/model/ma/Coach.java:7]:
```suggestion
@PrimaryKeyJoinColumn(name = "coach_id")
```

File: src/main/java/core/basesyntax/model/ma/Mentor.java [hunk 5]
```diff
@@ -1,5 +1,10 @@
 package core.basesyntax.model.ma;
 
+import javax.persistence.Entity;
+import javax.persistence.PrimaryKeyJoinColumn;
+
+@Entity
+@PrimaryKeyJoinColumn(name = "mentorId")
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/model/ma/Mentor.java:7]:
```suggestion
@PrimaryKeyJoinColumn(name = "mentor_id")
```

File: src/main/java/core/basesyntax/model/machine/Machine.java [hunk 6]
```diff
@@ -1,9 +1,27 @@
...
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Inheritance;
+import javax.persistence.InheritanceType;
+
+@Entity
+@Inheritance(strategy = InheritanceType.TABLE_PER_CLASS)
 public class Machine {
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/model/machine/Machine.java:12]: with this inheritance strategy we need a different id generation type
also, `@ID` annotation is missing

File: src/main/java/core/basesyntax/model/zoo/Animal.java [hunk 7]
```diff
@@ -1,9 +1,31 @@
...
+import javax.persistence.GenerationType;
+import javax.persistence.Inheritance;
+import javax.persistence.InheritanceType;
+
+@Entity
+@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
+@DiscriminatorColumn(name = "animal_type",
+        discriminatorType = DiscriminatorType.STRING)
 public class Animal {
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/model/zoo/Animal.java:16]: `@ID` annotation is missing

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/dao/animal/AnimalDaoImpl.java (hunk 1 above)
Comment (DariiaPikul) [reply src/main/java/core/basesyntax/dao/animal/AnimalDaoImpl.java:41]: Sorry, I thought it was a findBuColor method 🤦‍♀️

File: src/main/java/core/basesyntax/dao/ma/MentorDaoImpl.java (hunk 2 above)
Comment (AndrewPolishchuc) [reply src/main/java/core/basesyntax/dao/ma/MentorDaoImpl.java:8]: Sure. Done

File: src/main/java/core/basesyntax/dao/machine/MachineDaoImpl.java (hunk 3 above)
Comment (AndrewPolishchuc) [reply src/main/java/core/basesyntax/dao/machine/MachineDaoImpl.java:42]: Done

File: src/main/java/core/basesyntax/model/ma/Coach.java (hunk 4 above)
Comment (AndrewPolishchuc) [reply src/main/java/core/basesyntax/model/ma/Coach.java:7]: Done

File: src/main/java/core/basesyntax/model/ma/Mentor.java (hunk 5 above)
Comment (AndrewPolishchuc) [reply src/main/java/core/basesyntax/model/ma/Mentor.java:7]: Done

File: src/main/java/core/basesyntax/model/zoo/Animal.java (hunk 7 above)
Comment (AndrewPolishchuc) [reply src/main/java/core/basesyntax/model/zoo/Animal.java:16]: Added

File: src/main/java/core/basesyntax/model/machine/Machine.java (hunk 6 above)
Comment (AndrewPolishchuc) [reply src/main/java/core/basesyntax/model/machine/Machine.java:12]: Id added. Straregy shanges on "TABLE"

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request closed without merging**